l10n_it_delivery_note/models/sale_order.py

```diff
@@ -121,13 +124,138 @@ def _generate_delivery_note_lines(self, invoice_ids):
+            invoice_grouping_keys = self._get_invoice_grouping_keys()
+            invoice_vals_list = sorted(
+                invoice_vals_list,
+                key=lambda x: [
+                    x.get(grouping_key) for grouping_key in invoice_grouping_keys
+                ],
+            )
+            for _grouping_keys, invoices in groupby(
+                invoice_vals_list,
+                key=lambda x: [
+                    x.get(grouping_key) for grouping_key in invoice_grouping_keys
+                ],
+            ):
+                origins = set()
+                payment_refs = set()
+                refs = set()
+                ref_invoice_vals = None
+                for invoice_vals in invoices:
+                    if not ref_invoice_vals:
+                        ref_invoice_vals = invoice_vals
+                    else:
+                        ref_invoice_vals["invoice_line_ids"] += invoice_vals[
+                            "invoice_line_ids"
+                        ]
+                    origins.add(invoice_vals["invoice_origin"])
+                    payment_refs.add(invoice_vals["payment_reference"])
+                    refs.add(invoice_vals["ref"])
+                ref_invoice_vals.update(
+                    {
+                        "ref": ", ".join(refs)[:2000],
+                        "invoice_origin": ", ".join(origins),
+                        "payment_reference": len(payment_refs) == 1
+                        and payment_refs.pop()
+                        or False,
+                    }
+                )
+                new_invoice_vals_list.append(ref_invoice_vals)
+            invoice_vals_list = new_invoice_vals_list
+        if len(invoice_vals_list) < len(self):
+            SaleOrderLine = self.env["sale.order.line"]
+            for invoice in invoice_vals_list:
+                sequence = 1
+                for line in invoice["invoice_line_ids"]:
+                    line[2]["sequence"] = SaleOrderLine._get_invoice_line_sequence(
+                        new=sequence, old=line[2]["sequence"]
+                    )
+                    sequence += 1
+
+        moves = (
+            self.env["account.move"]
+            .sudo()
+            .with_context(default_move_type="out_invoice")
+            .create(invoice_vals_list)
+        )
+        if final:
+            moves.sudo().filtered(
+                lambda m: m.amount_total < 0
+            ).action_switch_invoice_into_refund_credit_note()
+        for move in moves:
+            move.message_post_with_view(
+                "mail.message_origin_link",
+                values={"self": move, "origin": move.line_ids.sale_line_ids.order_id},
+                subtype_id=self.env["ir.model.data"]._xmlid_to_res_id("mail.mt_note"),
+            )
+        # return moves
+        self._assign_delivery_notes_invoices(moves.ids)
+        self._generate_delivery_note_lines(moves.ids)
+        return moves
```

Comment:
**issue**: questa implementazione non chiama mai `super()`, rompendo tutti gli altri moduli che personalizzano `_create_invoices()`, tra cui i moduli core `sale_timesheet` e `repair` e svariati moduli OCA

Reply:
Potrebbe andare bene una soluzione del genere?
Chiamare la `super()` nel caso in cui ci sia attivata l'opzione è quasi impossibile perchè andrebbe fatta a priori per poi distruggere tutto quello che ha fatto per sostituire le righe con il nuovo raggruppamento.

L'unica cosa è che diventerebbe impossibile richiamare il metodo originale nel caso il flag non fosse attivo ed ho dovuto riportare il metodo originale e non mi piace molto...

Reply:
Con un monkeypatching di questo tipo, credo non bisognerebbe fare un nuovo model ma solo definire la funzione (fuori dalla classe), e poi andarla a sostituire alla funzione originale come hai fatto in fondo.

A mali estremi, potrebbe funzionare. Di certo la funzione originale è troppo lunga per andarsi a inserire in maniera efficace.

Abbiamo provato un'implementazione alternativa per la stessa idea di fondo (1 riga DDT = 1 riga fattura) che speriamo essere più compatibile qui https://github.com/OCA/l10n-italy/pull/4324